### PR TITLE
added apkdownload url to snack session

### PR DESCRIPTION
**Pull request by sierrakn**

### CODE REVIEW

File: src/SnackSession.js [hunk 1]
```diff
@@ -16,13 +16,15 @@ import difference from 'lodash/difference';
...
 import validate from 'validate-npm-package-name';
+import fp from 'lodash/fp';
 
 import * as DevSession from './utils/DevSession';
 import { defaultSDKVersion, sdkSupportsFeature } from './configs/sdkVersions';
 import preloadedModules from './configs/preloadedModules';
 import constructExperienceURL from './utils/constructExperienceURL';
 import sendFileUtils from './utils/sendFileUtils';
 import isModulePreloaded from './utils/isModulePreloaded';
+import buildAdkAsync from './utils/buildProject';
```
Comment (tcdavis) [inline src/SnackSession.js]: buildApk ?

File: src/utils/buildProject.js [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+  if (options.mode !== 'status' && (options.platform === 'android' || options.platform === 'all')) {
+    if (!exp.android || !exp.android.package) {
+      // todo: error handling
+      // Must specify a java package in order to build this experience for Android.
+      // Please specify one in ${configName} at "${configPrefix}android.package"
+    }
+  }
+
+  return await Api.callMethodAsync('build', [], 'put', {
```
Comment (tcdavis) [inline src/utils/buildProject.js]: Where is Api defined?

File: src/SnackSession.js [hunk 3]
```diff
@@ -292,6 +300,94 @@ export default class SnackSession {
...
+        },
+      },
+    };
+    return appJsonTemplate;
+  };
+
+  /**
+   * Builds an apk from the snack and returns a url to download the apk.
+   * @param 
+   * @returns {Promise.<void>} A promise that contains the url when fulfilled.
```
Comment (tcdavis) [inline src/SnackSession.js]: Promise<string> or whatever?

File: src/SnackSession.js [hunk 4]
```diff
@@ -292,6 +300,94 @@ export default class SnackSession {
...
+    const { id: buildId } = await buildAdkAsync(appJson, opts);
+    const completedJob = await this.wait(buildId, appJson, {});
+    const artifactUrl = completedJob.artifactId
+        ? `https://expo.io/artifacts/${completedJob.artifactId}`
+        : completedJob.artifacts.url;
+
+    return artifactUrl;
+  };
+
+  async wait(buildId, appJson, { timeout = 1200, interval = 60 } = {}) {
```
Comment (tcdavis) [inline src/SnackSession.js]: Should give wait a more descriptive name & have it end in async.  Prefix things that aren't intended to be called directly w/ _

My expectation would be that timeout / interval would be specified in milliseconds

File: src/SnackSession.js [hunk 5]
```diff
@@ -292,6 +300,94 @@ export default class SnackSession {
...
+      const res = await buildAdkAsync(appJson, {
+        current: false,
+        mode: 'status'
+      });
+      const job = fp.compose(
+        fp.head,
+        fp.filter(job => buildId && job.id === buildId),
+        fp.getOr([], 'jobs')
+      )(res);
+      switch (job.status) {
```
Comment (tcdavis) [inline src/SnackSession.js]: can this just be an if

File: src/SnackSession.js [hunk 6]
```diff
@@ -292,6 +300,94 @@ export default class SnackSession {
...
+  async wait(buildId, appJson, { timeout = 1200, interval = 60 } = {}) {
+    let time = new Date().getTime();
+    await sleep(secondsToMilliseconds(interval));
+    const endTime = time + secondsToMilliseconds(timeout);
+    while (time <= endTime) {
+      const res = await buildAdkAsync(appJson, {
+        current: false,
+        mode: 'status'
+      });
+      const job = fp.compose(
```
Comment (tcdavis) [inline src/SnackSession.js]: Pulling in fp stuff feels like overkill for this.  At least within js, I think the tradeoffs in clarity aren't worth it until you start wanting to have functions that return functions

What happens if buildId is something falsey?
Is it possible to have multiple jobs that make it past the filter condition?

File: src/SnackSession.js [hunk 7]
```diff
@@ -292,6 +300,94 @@ export default class SnackSession {
...
+    return appJsonTemplate;
+  };
+
+  /**
+   * Builds an apk from the snack and returns a url to download the apk.
+   * @param 
+   * @returns {Promise.<void>} A promise that contains the url when fulfilled.
+   * @function
+   */
+  getADKUrlAsync = async (
```
Comment (tcdavis) [inline src/SnackSession.js]: be consistent about casing.  I'd prefer getAdkUrlAsync

File: src/SnackSession.js [hunk 8]
```diff
@@ -292,6 +300,94 @@ export default class SnackSession {
...
+      platform?: string,
+      expIds?: Array<string>,
+      type?: string,
+      releaseChannel?: string,
+      bundleIdentifier?: string,
+      publicUrl?: string,
+      sdkVersion?: string,
+    } = {}
+  ): Promise<string> => {
+    opts.sdkVersion = this.sdkVersion;
```
Comment (tcdavis) [inline src/SnackSession.js]: Why accept these as arguments?

File: src/SnackSession.js [hunk 9]
```diff
@@ -292,6 +300,94 @@ export default class SnackSession {
...
+  /**
+   * Builds an apk from the snack and returns a url to download the apk.
+   * @param 
+   * @returns {Promise.<void>} A promise that contains the url when fulfilled.
+   * @function
+   */
+  getADKUrlAsync = async (
+    appJson : Object,
+    opts: {
+      current?: boolean,
```
Comment (tcdavis) [inline src/SnackSession.js]: isCurrent ? shouldMakeCurrent ?
It's not obvious to me what current, mode, expIds, or type mean or what valid values would look like

File: src/utils/buildProject.js [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+    expIds?: Array<string>,
+    type?: string,
+    releaseChannel?: string,
+    bundleIdentifier?: string,
+    publicUrl?: string,
+    sdkVersion?: string,
+  } = {}
+) {
+
+  const schema = joi.object().keys({
```
Comment (tcdavis) [inline src/utils/buildProject.js]: I believe the joi validation is there to provide users immediate and explicit feedback on project configuration issues.  Would any of these be supplied by end users working on snack, or just whatever projects are using the sdk?

Comment (sierrakn) [reply src/utils/buildProject.js]: (for now)

File: src/SnackSession.js (hunk 9 above)
Comment (sierrakn) [reply src/SnackSession.js]: removed.

File: src/SnackSession.js (hunk 8 above)
Comment (sierrakn) [reply src/SnackSession.js]: Yeah I think they're actually unneccesary for this.

Comment (tcdavis) [reply src/SnackSession.js]: they need to be on the opts you pass to buildAdkAsync, but it looks like any values you passed in to getADKUrlAsync would always be overwritten

File: src/utils/buildProject.js [hunk 11]
```diff
@@ -1,47 +1,67 @@
...
-    expIds: joi.array(),
-    type: joi.any().valid('archive', 'simulator', 'client'),
-    releaseChannel: joi.string().regex(/[a-z\d][a-z\d._-]*/),
-    bundleIdentifier: joi.string().regex(/^[a-zA-Z][a-zA-Z0-9\-\.]+$/),
-    publicUrl: joi.string(),
-    sdkVersion: joi.strict(),
+  options.platform = 'android';
+  exp = appJson.expo;
+
+  return await callMethodAsync('build', 'put', {
```
Comment (tcdavis) [inline src/utils/buildProject.js]: _callMethodAsync ?

File: src/SnackSession.js (hunk 3 above)
Comment (sierrakn) [reply src/SnackSession.js]: done.

**Pull request closed without merging**